packages/dds/tree/src/feature-libraries/sequence-field/markQueue.ts

```diff
@@ -0,0 +1,110 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/common-utils";
+import { RevisionTag } from "../../core";
+import { IdAllocator } from "../modular-schema";
+import { Mark, SizedMark } from "./format";
+import {
+    applyMoveEffectsToMark,
+    MoveEffectTable,
+    splitMarkOnInput,
+    splitMarkOnOutput,
+} from "./moveEffectTable";
+import { getInputLength, getOutputLength, isSizedMark } from "./utils";
+
+export class MarkQueue<T> {
+    private readonly stack: Mark<T>[] = [];
+    private index = 0;
+
+    public constructor(
+        private readonly list: readonly Mark<T>[],
+        public readonly revision: RevisionTag | undefined,
+        private readonly moveEffects: MoveEffectTable<T>,
+        private readonly genId: IdAllocator,
+        private readonly reassignMoveIds: boolean = false,
+        private readonly composeChanges?: (a: T | undefined, b: T | undefined) => T | undefined,
+    ) {
+        this.list = list;
+    }
+
+    public isEmpty(): boolean {
+        return this.peek() === undefined;
+    }
+
+    public splitInputAndDequeue(length: number): SizedMark<T> | undefined {
```

Comment:
Nit: I think `dequeueInput` would be a sufficient name, but I don't care too much.
Same deal with `splitOutputAndDequeue` => `dequeueOutput`.